File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    private boolean helper(int idx, int[] arr, int n, int target, Boolean[][] dp){
        // base cases
        if(idx >= n){
            return false;
        }
        if(target<0){
            return false;
        }
        if(target==0){
            return true;
        }
        if(dp[idx][target] != null){ // step 3
            return dp[idx][target]; 
        }

        boolean pick = helper(idx + 1, arr, n, target - arr[idx],dp);
        boolean notPick = helper(idx + 1, arr, n, target,dp);

        return dp[idx][target] = pick || notPick; // step 2
    }
    public boolean canPartition(int[] nums) {
        int N = nums.length;
        int total = 0;
        for(int i = 0;i<N;i++){
            total = total + nums[i];
        }
        if(total % 2 !=0){
            return false;
        }
        int target = total / 2;
        Boolean[][] dp = new Boolean[N][target+1]; // step 1
        return helper(0, nums, N, target,dp);
    }
}

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def solve(self, nums, target):
        n = len(nums)
        dp = [[False] * (target + 1) for _ in range(n) ]

        for i in range(n):
            dp[i][0] = True

        if nums[0] <= target:
            dp[0][nums[0]] = True
        
        for i in range(1,n):
            for t in range(1, target+1):
                incl = False
                if nums[i] < t:
                    incl = dp[i- 1][t - nums[i]]
                excl = dp[i - 1][t - 0]
                dp[i][t] = incl or excl

        return dp[n-1][target]

    def canPartition(self, nums: List[int]) -> bool:
        if (sum(nums) % 2 != 0):
            return False
        n = len(nums)
        target = sum(nums) // 2
        
        return self.solve(nums, target)